src/shogun/kernel/PolyKernel.cpp

```diff
@@ -71,8 +72,28 @@ void CPolyKernel::init()
 	m_gamma = 1.0;
 
 	set_normalizer(new CSqrtDiagKernelNormalizer());
-	SG_ADD(&degree, "degree", "Degree of polynomial kernel", ParameterProperties::HYPER);
-	SG_ADD(&m_c, "c", "The kernel is inhomogeneous if the value is higher than 0", ParameterProperties::HYPER);
-	SG_ADD(&m_gamma, "gamma", "Scaler for the dot product", ParameterProperties::HYPER);
+	SG_ADD(
+	    &degree, "degree", "Degree of polynomial kernel",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_c, "c", "The kernel is inhomogeneous if the value is higher than 0",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_gamma, "gamma", "Scaler for the dot product",
+	    ParameterProperties::HYPER | ParameterProperties::AUTO, [this]() {
+		    if (lhs->get_feature_class() >= EFeatureClass::C_STREAMING_DENSE &&
```

Comment:
I thought about this a bit more. Would it be possible to make this a normal function that is defined somewhere else (inside the same file)?
I think putting the lambda here is a bit messy, as then we mix registration and computation of auto values, where those probably are better separated in the code or?

Reply:
also we could avoid the clunky formatting?

Reply:
Another reason (and suggestion): it would be good to have `poly_kernel.put("degree", auto_param("num_feats")` as mentioned in my other comment. For that we would need

* the function being isolated somewhere so it can be accessed
* have an `auto_param` factory so that users can instantiate those strategies, which results in marking the parameter as auto as you did here. So the auto parameter function would need to be registered somewhere so that it is available to users.
* this means that there needs to be a base type to represent all auto strategies that could be used
* if I get a string representation of the parameter, say in Python, it would write `PolyKernel(degree=auto_param("bla"))`
* overload `put` to accept the type that is returned by the "auto_param" factory, and then register the instantiated function/pointer

Obviously, this is some work. Although it is not that bad. Doesnt all need to happen in this PR, but you see now why I don't like the lambda in there too much? Shall we discuss it a bit in here?

Reply:
OK, for this PR I can write the functions in a separate file. I could write a function wrapper that takes the references, i.e. the feature matrix, and then returns a lambda that has this captured and that can be called when the algorithm is initialised?
And then in another PR we can write the put overloads using those functions? I mean the principle of your suggestion and what I wrote is similar right? In both ways we need a way to register that a parameter has to be initialised in the future.

Reply:
Yes (no need for a separate file though as the auto strategy can be part of this class/plugin)
I don’t really understand why a lambda is necessary, why not just a function but that is a detail I guess
My main point is that things need to be structured in a way that later on one can easily add user access to change those strategies and documentation